src/Lean/Widget/UserWidget.lean

```diff
@@ -0,0 +1,189 @@
+  deriving Inhabited, ToJson, FromJson
+
+/-- Use this structure and the `@[widget]` attribute to define your own widgets.
+
+```lean
+@[widget]
+def rubiks : UserWidgetDefinition :=
+  { name := "Rubiks cube app"
+    javascript := include_str ...
+  }
+```
+-/
+structure UserWidgetDefinition where
+  /-- Pretty name of user widget to display to the user. -/
+  name : String
+  /-- An ESmodule that exports a react component to render. -/
+  javascript: String
+  deriving Inhabited, ToJson, FromJson
+
+structure UserWidget where
+  id : Name
+  /-- Pretty name of widget to display to the user.-/
+  name : String
+  javascriptHash: UInt64
+  deriving Inhabited, ToJson, FromJson
+
+private abbrev WidgetSourceRegistry := SimplePersistentEnvExtension
+    (UInt64 × WidgetSource)
+    (Std.RBMap UInt64 WidgetSource compare)
+
+-- Mapping widgetSourceId to hash of sourcetext
+builtin_initialize userWidgetRegistry : MapDeclarationExtension UserWidget ← mkMapDeclarationExtension `widgetRegistry
+builtin_initialize widgetSourceRegistry : WidgetSourceRegistry ←
+  registerSimplePersistentEnvExtension {
+    name          := `widgetSourceRegistry
+    addImportedFn := fun xss => xss.foldl (Array.foldl (fun s n => s.insert n.1 n.2)) ∅
+    addEntryFn    := fun s n => s.insert n.1 n.2
+    toArrayFn     := fun es => es.toArray
+  }
+
+private unsafe def attributeImplUnsafe : AttributeImpl where
+  name := `widget
+  descr := "Mark a string as static code that can be loaded by a widget handler."
+  applicationTime := AttributeApplicationTime.afterCompilation
+  add decl _stx _kind := do
+    let env ← getEnv
+    let defn ← evalConstCheck UserWidgetDefinition ``UserWidgetDefinition decl
+    let javascriptHash := hash defn.javascript
+    let env := userWidgetRegistry.insert env decl {id := decl, name := defn.name, javascriptHash}
+    let env := widgetSourceRegistry.addEntry env (javascriptHash, {sourcetext := defn.javascript})
+    setEnv <| env
+
+@[implementedBy attributeImplUnsafe]
+opaque attributeImpl : AttributeImpl
+
+builtin_initialize registerBuiltinAttribute attributeImpl
+
+/-- Input for `getWidgetSource` RPC. -/
+structure GetWidgetSourceParams where
+  /-- The hash of the sourcetext to retrieve. -/
+  hash: UInt64
+  pos : Lean.Lsp.TextDocumentPositionParams
+  deriving ToJson, FromJson
+
+open Lean.Server Lean
+
+open RequestM in
+@[serverRpcMethod]
+def getWidgetSource (args : GetWidgetSourceParams) : RequestM (RequestTask WidgetSource) :=
+  RequestM.withWaitFindSnapAtPos args.pos fun snap => do
+    let env := snap.cmdState.env
+    if let some w := widgetSourceRegistry.getState env |>.find? args.hash then
+      return w
+    else
+      throw <| RequestError.mk .invalidParams s!"No registered user-widget with hash {args.hash}"
+
+open Lean Elab
+
+/--
+  Try to retrieve the `UserWidgetInfo` at a particular position.
+-/
+def widgetInfoAt? (text : FileMap) (t : InfoTree) (hoverPos : String.Pos) : List UserWidgetInfo :=
```

Comment:
Should rather be `widgetInfosAt`?

Reply:
@EdAyers Apologies for being so nitpicky, but there should not be a `?`. The docstring should be adjusted as well.